### DISCUSSION TITLE
Nameserver and http proxy of renovate/node in dind-setup

### DISCUSSION THREAD

**Discussion by chkpnt** [renovatebot/renovate#18422] [Request Help] [answered]
I'm trying to switch to a Docker-in-Docker (dind) setup for my on-prem GitLab installation.

The container, which is starting the image `renovate/node` as `renovate_node`, uses the correct docker-internal nameserver:

```
ubuntu@runner-urk1hanh-project-7-concurrent-0:/usr/src/app$ cat /etc/resolv.conf
search <redacted>
nameserver 127.0.0.11
options ndots:0
```

The env variables `http_proxy`, `https_proxy` and `no_proxy` (which have to include "docker") are also set correctly, as configured in the runner's configuration. Therefore, the docker image `renovate/node` can be loaded successfully.

But within the container `renovate_node`, the nameserver is configured to the docker defaults:

```
user@d4ffcb0ad5f7:/builds/renovate/renovate-runner/renovate/repos/gitlab/test-project$ cat /etc/resolv.conf
search <redacted>
options ndots:0

nameserver 8.8.8.8
nameserver 8.8.4.4
```

Furthermore, the env variables `http_proxy`, `https_proxy` and `no_proxy` aren't set. Therefore, the container itself has no working network and the command `dumb-init -- bash -l -c install-tool npm 8.19.2 && hash -d npm 2>/dev/null || true && npm install --package-lock-only --no-audit --ignore-scripts` fails with a timeout.

What am I missing?

**Comment by viceice** [comment]
that's probably an issue with your dind configuration. renovate simply uses docker cli to start the sidecar container. renovate will also pass the proxy ENV vars to sidecar container automatically. check the debug logs, there should be the full command logged

**Reply by rarkins** [reply]
Double check upper lower case too

**Reply by chkpnt** [reply]
I've just changed the Variables from lower case to upper case, but they still aren't passed to the sidecar container.

```
[[runners]]
  name = "Docker Runner for Renovate (dind)"
  url = "https://gitlab-test.redacted.de"
  id = 10
  token = "..."
  token_obtained_at = 2022-10-20T07:23:15Z
  token_expires_at = 0001-01-01T00:00:00Z
  executor = "docker"
  environment = ["HTTP_PROXY=http://proxy.redacted.de:3128", "HTTPS_PROXY=http://proxy.redacted.de:3128", "NO_PROXY=localhost,127.0.0.1,.redacted.de,docker"]
  [runners.custom_build_dir]
  [runners.cache]
    [runners.cache.s3]
    [runners.cache.gcs]
    [runners.cache.azure]
  [runners.docker]
    tls_verify = false
    image = "docker:20.10.16"
    privileged = true
    disable_entrypoint_overwrite = false
    oom_kill_disable = false
    disable_cache = false
    volumes = ["/certs/client", "/cache"]
    network_mode = "gateway"
    shm_size = 0
    helper_image_flavor = "ubuntu"
```

Gateway is the name of my bridge network.

Do you see any problems with my config?

**Reply by chkpnt** [reply]
It seems the container `renovate_node` is started with the wrong network:

```
ubuntu@runner-urk1hanh-project-7-concurrent-0:/usr/src/app$ docker container inspect renovate_node
[...]
            "Networks": {
                "bridge": {
                    "IPAMConfig": null,
                    "Links": null,
                    "Aliases": null,
                    "NetworkID": "0c67658d5df6d1e98b031c641ba43cf14f543d66b2489a20e952b7fcf4c125b0",
                    "EndpointID": "623102aa42cd9d0fd1f63523e8ddc106fb89f7ea69d2538cbbff9cccc376759b",
                    "Gateway": "172.17.0.1",
                    "IPAddress": "172.17.0.2",
                    "IPPrefixLen": 16,
                    "IPv6Gateway": "",
                    "GlobalIPv6Address": "",
                    "GlobalIPv6PrefixLen": 0,
                    "MacAddress": "02:42:ac:11:00:02",
                    "DriverOpts": null
                }
            }
[...]
```

This should be `gateway` instead.

**Reply by chkpnt** [reply]
Mh, the dind-container doesn't know the network:

```
ubuntu@runner-urk1hanh-project-7-concurrent-0:/usr/src/app$ docker network ls
NETWORK ID     NAME      DRIVER    SCOPE
0c67658d5df6   bridge    bridge    local
ba4a7fcf9a43   host      host      local
1ea03186b5a4   none      null      local
```

This are the networks available on the host:
```
ltksle15:~ # docker network ls
NETWORK ID     NAME                DRIVER    SCOPE
e901e6576024   bridge              bridge    local
6063404045e8   gateway             bridge    local <- this should be used
a1c3d6668fd9   gitlab_backend      bridge    local
1e0887437862   host                host      local
c861af63453d   none                null      local
490ed1e8d669   sonarqube_backend   bridge    local
```

The dind-container itself is using the gateway-network:
```
> docker container inspect 9614f93c9c5b
[...]
            "Networks": {
                "gateway": {
                    "IPAMConfig": null,
                    "Links": null,
                    "Aliases": [
                        "build",
                        "runner-urk1hanh-project-7-concurrent-0-61b407a042fd004e-build-3",
                        "9614f93c9c5b",
                        "runner-urk1hanh-project-7-concurrent-0"
                    ],
                    "NetworkID": "6063404045e832622fa53c57652b7db5168b1a0b8d3bdc9bfbba139a8f4a0696",
                    "EndpointID": "852989968d0e8af9bb57926165f260c81c4cebbb93ef664347d1327165d42dab",
                    "Gateway": "172.16.1.1",
                    "IPAddress": "172.16.1.8",
                    "IPPrefixLen": 24,
                    "IPv6Gateway": "",
                    "GlobalIPv6Address": "",
                    "GlobalIPv6PrefixLen": 0,
                    "MacAddress": "02:42:ac:10:01:08",
                    "DriverOpts": null
                }
            }
[...]
```

**Reply by chkpnt** [reply]
Okay, I figured it out: The container within dind needs to be started with `--network host`, so that the internal nameserver is used. But how can I configure renovate to do so?

>   renovate will also pass the proxy ENV vars to sidecar container automatically. 

Unfortunately, this isn't the case. So how can I enforce it?

**Comment by chkpnt** [comment, marked as answer]
After switching from a service based dind setup to mounting `docker.sock`, the nameserver and http proxy issues were gone. So this seems to be the answer.

But I run into other issues... :-(

>  "message": "Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/builds/dschung/renovate-runner/renovate/repos/gitlab/test-project\":\"/builds/dschung/renovate-runner/renovate/repos/gitlab/test-project\" -v \"/builds/dschung/renovate-runner/renovate/cache\":\"/builds/dschung/renovate-runner/renovate/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/builds/dschung/renovate-runner/renovate/repos/gitlab/test-project\" docker.io/renovate/sidecar bash -l -c \"install-tool node v18.13.0 && install-tool npm 8.19.3 && hash -d npm 2>/dev/null || true && npm install --package-lock-only --no-audit --ignore-scripts\"\nYou don't have permission to create /builds/dschung/renovate-runner/renovate/cache/containerbase.\ncurl: (23) Failed writing received data to disk/application\nDownload failed: https://nodejs.org/dist/v18.13.0/node-v18.13.0-linux-x64.tar.xz\nbash: npm: command not found\n",

But doesn't matter any longer. I've noticed that trying to run renovate with a dind setup isn't needed at all, as `binarySource=install` works fine. 🤦‍♂️
